### PR TITLE
LinAlg impl in detail

### PR DESCRIPTION
**Pull request by divyegala**
Addresses #330

### CODE REVIEW

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: cpp/include/raft/distance/detail/distance.cuh [hunk 1]
```diff
@@ -31,6 +30,7 @@
 #include <raft/distance/detail/l1.cuh>
 #include <raft/distance/detail/minkowski.cuh>
 #include <raft/distance/detail/russell_rao.cuh>
+#include <raft/linalg/distance_type.hpp>
```
Comment (cjnolet) [inline cpp/include/raft/distance/detail/distance.cuh:33]: We should use this opportunity to move `distance_type` into `raft/distance`.

File: cpp/include/raft/handle.hpp [hunk 2]
```diff
@@ -32,10 +32,10 @@
 ///@todo: enable once we have migrated cuml-comms layer too
 //#include <common/cuml_comms_int.hpp>
 
-#include <raft/linalg/cublas_wrappers.h>
-#include <raft/linalg/cusolver_wrappers.h>
 #include <raft/sparse/cusparse_wrappers.h>
 #include <raft/comms/comms.hpp>
+#include <raft/linalg/cublas_wrappers.hpp>
```
Comment (cjnolet) [inline cpp/include/raft/handle.hpp:37]: These should be in detail. They aren't intended to be public and are only meant to be used as a facade for the actual cublas/cusolver/cusparse/etc... calls.

File: cpp/include/raft/linalg/binary_op.hpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/*
```
Comment (cjnolet) [inline cpp/include/raft/linalg/binary_op.hpp:1]: The years need to be updated everywhere.

File: cpp/include/raft/linalg/eig.hpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+           cudaStream_t stream)
+{
+  detail::eigDC(handle, in, n_rows, n_cols, eig_vectors, eig_vals, stream);
+}
+
+using detail::COPY_INPUT;
+using detail::EigVecMemUsage;
+using detail::OVERWRITE_INPUT;
+
+#if CUDART_VERSION >= 10010
```
Comment (cjnolet) [inline cpp/include/raft/linalg/eig.hpp:53]: We should be able to drop this now since RAPIDS is always >=CTK 11

**Review by cjnolet — CHANGES_REQUESTED** [review]
I've only gone through it quickly in a first pass so far but changes look great for the most part. Just saw a couple smaller things initially.

File: cpp/include/raft/handle.hpp (hunk 2 above)
Comment (divyegala) [reply cpp/include/raft/handle.hpp:37]: Oh yep, you are totally right - missed that one my bad

Comment (cjnolet) [reply cpp/include/raft/handle.hpp:37]: Jus to update this conversation so it's not lost over the holidays, the intention here is to wrap the standard cuda library functions so we centralize all the calls but to provide RAFT public API functions which are completely agnostic of the standard cuda libraries underneath.

As discussed in https://github.com/rapidsai/cuml/pull/4446, for a short time, we might have a few places in the cuml solver code which is calling down to raft's `linalg::detail` namespace. These will need to be updated to use the public API, once it's available (and we do that in a follow-on PR).

**Review by cjnolet — CHANGES_REQUESTED** [review]
This is a lot of files so I might need to do a second pass over them. So far, it's looking really good, though. The couple things I've found are very minor.

File: cpp/include/raft/spectral/spectral_util.hpp [hunk 5]
```diff
@@ -18,6 +18,7 @@
 
 #include <raft/cudart_utils.h>
 #include <raft/handle.hpp>
+#include <raft/linalg/detail/cublas_wrappers.hpp>
```
[the remark this answers is not in the document]
Comment (divyegala) [reply cpp/include/raft/spectral/spectral_util.hpp:21]: I have not worked with `spectral` in this PR. Should this not be addressed when we work with that package?

[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/include/raft/spectral/spectral_util.hpp:21]: Ah, yeah I overlooked the packaging during review. We can disregard this one for now.

File: cpp/include/raft/distance/distance_type.hpp [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, NVIDIA CORPORATION.
+ * Copyright (c) 2021, NVIDIA CORPORATION.
  *
```
Comment (cjnolet) [inline cpp/include/raft/distance/distance_type.hpp:3]: Stray copyright that needs updating

File: cpp/include/raft/linalg/detail/eig.hpp [hunk 7]
```diff
@@ -16,17 +16,20 @@
 
 #pragma once
 
+#include "cusolver_wrappers.hpp"
```
Comment (cjnolet) [inline cpp/include/raft/linalg/detail/eig.hpp:19]: I think this is duplicated.

File: cpp/include/raft/linalg/detail/lanczos.hpp [hunk 8]
```diff
@@ -0,0 +1,1400 @@
...
+
+  // -------------------------------------------------------
+  // Compute second Lanczos vector
+  // -------------------------------------------------------
+  if (*iter <= 0) {
+    *iter = 1;
+
+    // Apply matrix
+    if (shift != 0)
+      CUDA_TRY(cudaMemcpyAsync(lanczosVecs_dev + n,
```
Comment (cjnolet) [inline cpp/include/raft/linalg/detail/lanczos.hpp:126]: We need to be mindful that we update these as well while we're moving this stuff over. `RAFT_CUDA_TRY`, `RAFT_CUBLAS_TRY`, etc...

**Review by cjnolet — CHANGES_REQUESTED** [review]
Sorry for the delayed in the review- this is a massive set of changes to go through. Mostly small things at this point. This is almost ready to go. Will look through cuml side next.

File: cpp/include/raft/linalg/detail/map.cuh [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+template <typename InType, typename OutType, typename MapOp, int TPB, typename... Args>
+void mapImpl(
+  OutType* out, size_t len, MapOp map, cudaStream_t stream, const InType* in, Args... args)
+{
+  const int nblks = raft::ceildiv(len, (size_t)TPB);
+  mapKernel<InType, OutType, MapOp, TPB, Args...>
+    <<<nblks, TPB, 0, stream>>>(out, len, map, in, args...);
+  CUDA_CHECK(cudaPeekAtLastError());
```
Comment (cjnolet) [inline cpp/include/raft/linalg/detail/map.cuh:43]: Need another rename from `CUDA_CHECK` -> `RAFT_CUDA_TRY`

File: cpp/include/raft/linalg/detail/map_then_reduce.cuh [hunk 10]
```diff
@@ -88,70 +89,9 @@ void mapThenReduceImpl(OutType* out,
   const int nblks = raft::ceildiv(len, (size_t)TPB);
   mapThenReduceKernel<InType, OutType, MapOp, ReduceLambda, TPB, Args...>
     <<<nblks, TPB, 0, stream>>>(out, len, neutral, map, op, in, args...);
-  RAFT_CUDA_TRY(cudaPeekAtLastError());
+  CUDA_CHECK(cudaPeekAtLastError());
```
Comment (cjnolet) [inline cpp/include/raft/linalg/detail/map_then_reduce.cuh:92]: Here too

File: cpp/include/raft/linalg/detail/subtract.cuh [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+                       const math_t* singleScalarDev,
+                       IdxType len,
+                       cudaStream_t stream)
+{
+  // Just for the note - there is no way to express such operation with cuBLAS in effective way
+  // https://stackoverflow.com/questions/14051064/add-scalar-to-vector-in-blas-cublas-cuda
+  const IdxType nblks = raft::ceildiv(len, (IdxType)TPB);
+  subtract_dev_scalar_kernel<math_t>
+    <<<nblks, TPB, 0, stream>>>(outDev, inDev, singleScalarDev, len);
+  CUDA_CHECK(cudaPeekAtLastError());
```
Comment (cjnolet) [inline cpp/include/raft/linalg/detail/subtract.cuh:64]: Here too

**Review by cjnolet — APPROVED** [review]

**Review by cjnolet — APPROVED** [review]
LGTM

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**